# Lab 5 - Sudhan Shrestha 8889436

#### 1. Utilize the diabetes dataset from lab 4. Perform cross-validation on nine polynomial models, ranging from degree 0 to 8. (2 points)

In [11]:
# Importing the necessary packages
from sklearn import datasets
from sklearn.metrics import mean_absolute_error
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.pipeline import make_pipeline
from sklearn.model_selection import cross_val_score
from sklearn import metrics
import numpy as np
import pandas as pd

# importing the dataset used in lab 4
X, y = datasets.load_diabetes(as_frame=True, scaled=False, return_X_y=True)

#### 2. Construct a table summarizing the cross-validation results. Each model should have a separate row in the table. Include the R-Squared and Mean Absolute Error (MAE) metrics for each model. Calculate the mean value and standard deviation of these metrics from the cross-validation. Include both values. (2 points)

In [12]:
degrees = np.arange(0, 9)  # creating degree rangin from 0 to 8

cross_val_results = [] # empty list to store all crossvalidation metrics
# looping through every degree
for degree in degrees:
    # creating a pipeline. to create polynomial feature and apply linear regression
    model = make_pipeline(PolynomialFeatures(degree), LinearRegression())
    # using cross-validation to find the r-squared coefficient score across 5 splits of data
    r2_score = cross_val_score(model, X, y, cv=5, scoring='r2')
    r2_mean = r2_score.mean()
    r2_std = r2_score.std()
    # using cross-validation to find the negative-mean absoulte error score across 5 splits of data
    mae_score = cross_val_score(model, X, y, cv=5, scoring='neg_mean_absolute_error')
    mae_mean = mae_score.mean()
    mae_std = mae_score.std()
    # Stroing all these cross-validation score results.
    cross_val_results.append([r2_mean, r2_std, mae_mean, mae_std])

In [15]:
# Creating a data frame from the data generated above

columns = ['R-squared mean', 'R-squared standard devation', 'MAE mean', 'MAE standard devation']
metrics_df = pd.DataFrame(cross_val_results, columns=columns)
metrics_df

R-squared mean  R-squared standard devation     MAE mean   
0       -0.027506                     0.036772   -66.045624  \
1        0.482316                     0.049269   -44.276499   
2       -1.485182                     1.618007   -80.595439   
3     -203.418938                   225.878146  -342.051924   
4     -571.083108                   369.891884  -657.260477   
5     -436.856887                   379.100423  -562.993636   
6    -1694.550210                  2629.989561  -742.717488   
7    -5530.894075                  9518.586954 -1032.681725   
8   -16076.255118                 28049.952580 -1475.658532   

   MAE standard devation  
0               3.474660  
1               2.100110  
2              24.657166  
3             142.437752  
4             159.475902  
5              59.917202  
6             190.709474  
7             393.439617  
8             706.280274

#### 3. Identification of the Best Model: Identify the model that exhibits the highest performance based on the R-Squared and MAE metrics. Provide an explanation for choosing this specific model. (2 points)

In [14]:
# choosing the degree with the highest cross validation score on R-Squared
print("The best degree in both scoring metrics : ")
best_degree = degrees[np.argmax(metrics_df['R-squared mean'])]
print(f"R-Squared metrics: The best degree is {best_degree}.")
# choosing the degree with the highest cross validation score on MAE
best_degree = degrees[np.argmax(metrics_df['MAE mean'])]
print(f"MAE: The best degree is {best_degree}.")

The best degree in both scoring metrics : 
R-Squared metrics: The best degree is 1.
MAE: The best degree is 1.



To find the best model we need to determine the mean value from both the metrics R-Squarred and MAE which are used to find the best fit.

After observing the table above and perofming the `argmax()` function on both the R-Squared and MAE we come to find that the model with degree 1 has the highest R-squarted mean value among all the other values `(0.482316)`, which means that it has a better fit than other values as; A R<sup>2</sup>  of 1 indicates that the regression predictions perfectly fit the data and values of R<sup>2</sup> outside the range of 0 to 1 occurs when the model fit the data worse than the wrost possinle least-square predictor [Source- wiki](https://en.wikipedia.org/wiki/Coefficient_of_determination).

Also the Mean Absolute Error (MAE) for degree 1 is  `(-44.276499)` which means it has better fit; as MAE is an error metics which means the lower the value the better the fit and since we have negative MAE the opposite is true in our case.

Hence, since `degree of 1` has both the highest R-squared and the lowest Mean Absoulte Error (MAE) we can say that the `model with degree 1` has the best fit among all.